### ISSUE TITLE
Potential broken rand() in rdrand.h

### ISSUE THREAD

**Issue by by700git** [edenhill/librdkafka#2795]
Read the FAQ first: https://github.com/edenhill/librdkafka/wiki/FAQ

Description
===========
Not sure if anyone noticed this before, but seems this `rd_jitter` is used across multiple functions such as assigning partitions and select broker to send request:

in `rdrand.h`
```c
static RD_INLINE int rd_jitter (int low, int high) {
	return (low + (rand() % ((high-low)+1)));
	
}
```
I tested the behavior of `rand()` on multiple machines with same OS version (Windows Server 2016), it seems they are generating the same random sequence.

In our prod scenario, this potentially cause all newly started consumers send metadata request to the same broker, and the broker's request queue get filled up.

Potential fix might be:
1. Use something like `srand(time(NULL))`?
2. Or use `random(3)`? https://linux.die.net/man/3/rand

How to reproduce
================
In description

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): `master` `commit 71a9d8568edf3044af28756d22b44b7bcb1ac7ca` `v1.4.0-RC7`
 - [x] Apache Kafka version: `2.3.1`
 - [x] librdkafka client configuration: `all confs`
 - [x] Operating system: `Windows Server 2016`
 - [x] Provide logs `N/A`
 - [x] Provide broker log excerpts
 - [x] Critical issue: In our case, the broken rand() potentially cause consumers on different machines to send request to one broker, which will fill up the request queue if these are some heavy requests.

**Comment by edenhill** [comment]
We rely on the application initializing the random generator with a seed by calling srand(), but this isn't documented, which is clearly a flaw.

We could call srand()/srandom() ourselves, but I'm not completely certain that's a good idea for a library to do since it will affect the application's use of rand(), i.e., if an application relies on predictable random numbers by supplying its own seed with srand() we will silently break that by calling srand() internally.
One way around that would be to have a `enable.random.seed` which defaults to `true`, having librdkafka call srand() once on startup, but that the application can disable that by setting `enable.random.seed=false`.

Comments?

**Comment by rolandyoung** [comment]
Better to default to false, so that correctly written apps don't have to make any change? Just documenting the setting should provide a sufficient hint to those to whom this matters.

**Comment by edenhill** [comment]
I think the problem is that most applications don't rely on predictive randomness, so those devs will not see this property and thus not get proper randomess in librdkafka. So I think srand should be done by default to cater to the majority of users.
Now for the predictive randomess applications, with librdkafka calling rand() the predictiveness of rand() is screwed for the application anyway, so maybe it doesn't matter if we call srand(), but to be safe it should be opt-out-able

**Comment by rolandyoung** [comment]
Yes, thinking about it further, that is a fair comment. There is no harm in one additional call to srand (would that be one per rd_kafka_t handle, or managed by a global static?) except in cases where reproducible randomness is wanted.

**Comment by edenhill** [comment]
It should preferably just be once per runtime, so we can check it in rd_kafka_new() and keep track of if we've seeded it before.

This will not make v1.4.0.

**Comment by by700git** [comment]
Hi @edenhill and @rolandyoung Thanks for the comments! Another use case I can think of is, for other clients relying on librdkafka (dotnet for example), should they use a C# interface to set random seed (when writing C# applications), or should we set the seed in C# client's underlying implementation?

**Comment by edenhill** [comment]
C#'s  PRNG is probably not the same as the libc rand(), so setting a C# seed does not help.
https://docs.microsoft.com/en-us/dotnet/api/system.random?view=netframework-4.8#remarks

**Comment by by700git** [comment]
oh what I mean is, should we expose a librdkafka interface set_seed() for C# client to use? C# client doesn't seem to have a delegate to any seed related native methods now. BTW if fits, I can also take the responsibility for the PR. I take it as a good chance to get familiar with librdkafka :-)

**Comment by edenhill** [comment]
@by700git Didn't see this until now, we very much like contributions but this particular issue has been fixed in PR #2795, do you want me to find you an open issue that is a good entry level thing into librdkafka?

**Comment by edenhill** [comment]
Fixed in v1.4.2
